core/activity/src/dao/activity.rs

```diff
@@ -114,13 +114,23 @@ impl<'c> ActivityDao<'c> {
         .await
     }
 
-    pub async fn _get_activity_ids(&self) -> Result<Vec<String>> {
+    pub async fn get_activity_ids(&self, max_items: Option<u32>) -> Result<Vec<String>> {
         use schema::activity::dsl;
-        do_with_transaction(self.pool, "activity_dao_get_activity_ids", |conn| {
-            dsl::activity
-                .select(dsl::natural_id)
-                .get_results(conn)
-                .map_err(|e| e.into())
+
+        do_with_transaction(self.pool, "activity_dao_get_activity_ids", move |conn| {
+            if let Some(items) = max_items {
+                dsl::activity
+                    .order_by(dsl::id.desc())
+                    .limit(items.into())
+                    .select(dsl::natural_id)
+                    .get_results(conn)
+                    .map_err(|e| e.into())
+            } else {
+                dsl::activity
+                    .select(dsl::natural_id)
+                    .get_results(conn)
+                    .map_err(|e| e.into())
+            }
```

Comment:
Here is example how to add additional filters to query dynamically to get rid of duplication and avoid compilation problems: https://github.com/golemfactory/yagna/blob/dc16197f8037b404cea9d5ad97e0782ff21bff17/core/payment/src/dao/payment.rs#L227-L243
In short: you must use `.into_boxed()` on the first query